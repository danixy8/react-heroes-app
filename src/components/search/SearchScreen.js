
import { useNavigate, useLocation } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { getHeroesByName } from '../../selectors/getHeroesByName';
import { HeroCard } from '../hero/HeroCard';
import queryString from 'query-string';
import { useMemo } from 'react';

export const SearchScreen = () => {

  const navigate = useNavigate();
  const location = useLocation();

  const { q = '' } = queryString.parse(location.search);

  const [ formValues, handleInputChange ] = useForm( {
    searchText: q
  });

  const { searchText } = formValues;

  const heroesFilter = useMemo(() => getHeroesByName(q), [q]);

  const handleSearch = (e) => {

    console.log(searchText);

    e.preventDefault();
    navigate(`?q=${searchText}`)
  };

  return (
    <>
      <h1>Busquedas</h1>
      <hr />

      <div className="row">

        <div className="col-sm-12 col-md-5">
          <h4>Buscar Personaje</h4>
          <hr />

          <form onSubmit={ handleSearch }>
            <input 
              type="text" 
              placeholder="Buscar un heroe"
              className="form-control"
              name="searchText"
              autoComplete="off"
              onChange={ handleInputChange }
              value={ searchText }
            />
          
          <button 
            className="btn btn-outline-primary mt-2 btn-block"
            type="submit"
          >
            Buscar...
          </button>
          
          </form>


        </div>
 
          <div className="col-sm-12 col-md-7">
          { 
            q !== '' &&
            <>
              <h4>Resultados</h4> 
              <hr />
            </>
          }
          
          { heroesFilter.length > 0 && (
              heroesFilter.map(hero => (
                <HeroCard
                  key={ hero.id }
                  {...hero}
                />
              ))
           )
          }

          { 
            q !== '' && heroesFilter.length === 0 
            ?
              <p className="alert alert-danger">
                { `Personaje ${q} no encontrado`}
              </p>
            :
            null
          }

          </div>

      </div>
    </>
  )
}