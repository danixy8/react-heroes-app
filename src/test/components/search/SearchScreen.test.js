import { mount } from "enzyme";
import { MemoryRouter } from "react-router";
import { SearchScreen } from "../../../components/search/SearchScreen";

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate
}));

describe('Pruebas en <SearchScreen/>', () => {
  
  test('debe de mostrarse correctamente con valores por defecto', () => {
  
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <SearchScreen />
      </MemoryRouter>
    );

    expect( wrapper ).toMatchSnapshot();
    expect(wrapper.find('.btn').text().trim()).toBe('Buscar...');

  });

  test('debe de mostrar a Batman y el input con el valor del queryString', () => {
    
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <SearchScreen />
      </MemoryRouter>
    );

    expect(wrapper.find('input').prop('value')).toBe('batman');
    expect(wrapper.find('.card-title').text().trim()).toBe('Batman');

  });

  test('debe de ocultar el column -busqueda- si no hay valores', () => {
    
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=']}>
        <SearchScreen />
      </MemoryRouter>
    );

    expect(wrapper.find('h4').text().trim()).not.toBe('Resultados');
    
  });

  test('debe de mostrar un error si no se encuentra un heroe', () => {
    
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=joker123']}>
        <SearchScreen />
      </MemoryRouter>
    );

    expect(wrapper.find('.alert-danger').text().trim()).toBe('Personaje joker123 no encontrado');

  });

  test('debe de llamar el navigate a la nueva pantalla', () => {

    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <SearchScreen />
      </MemoryRouter>
    );

    wrapper.find('input').simulate('change', {
      target: { 
        name: 'searchText',
        value: 'batman'
      }
    });

    wrapper.find('form').prop('onSubmit')({
      preventDefault: () => {}
    });
    
    expect( mockNavigate ).toHaveBeenCalledWith('?q=batman');

  });

});