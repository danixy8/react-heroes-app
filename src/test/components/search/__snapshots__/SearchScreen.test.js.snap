// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen/> debe de mostrarse correctamente con valores por defecto 1`] = `
Array [
  <h1>
    Busquedas
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-sm-12 col-md-5"
    >
      <h4>
        Buscar Personaje
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Buscar un heroe"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-2 btn-block"
          type="submit"
        >
          Buscar...
        </button>
      </form>
    </div>
    <div
      className="col-sm-12 col-md-7"
    />
  </div>,
]
`;
